### DISCUSSION TITLE
Control non-standard and edge-case lights?

### DISCUSSION THREAD

**Discussion by Joshfindit** [monty68/uniled#41] [Ideas]
I have a couple lights that might be interesting, but it's proven important to set the scope of a project before dev time gets overwhelmed with edge cases.

Like: "Smart fairy lights" (3-wire WS2812BRGBIC with a built-in USB controller. You can get them on sale for $6 for 10M). The USB controller can be controlled over bluetooth with the iDeal LED app.

Or one of the matrix displays that support BT control (the one I have in-hand is 32x32 and controlled with the iDotMatrix app. It has an additional quirk: to show the correct time it must receive the time via BT after it has power)

I'm sure the number and diversity of addressable LED devices with BT controllers will only increase. If this is a direction you want to go it would probably be worth planning a plugin architecture so that coding/submitting new controllers is easier and reviewing the PRs is easier.

**Comment by monty68** [comment]
Hi Josh, I always intended to expand UniLED to cover many of the edge cases most others ignore, hence the name. One of the reasons I launched into the v2 development was that when adding some of the newer SP63X devices, I realised my framework was
not as flexible as I'd hoped.  Once familiar with the UniLED way, it is pretty quick to add a device when you have a dump and an understanding of the protocol and all the parameters and names used within its UI/app.

If I ever get a chance to write up some developer notes, it would make it easier to submit PRs, but for now, if people have the time to capture the protocol traffic and provide all the information, I can quickly add new devices in. But at the moment I have to buy each of the devices which can be fraught, only this month I had ordered the SP602E and SP608E devices and got scammed losing my money. OK not a huge amount but still GBP£50 down the drain.
